File: packages/amplify_auth_cognito/ios/Classes/FlutterUpdateUserAttributesRequest.swift

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import Foundation
import Amplify
import AmplifyPlugins
import amplify_core

struct FlutterUpdateUserAttributesRequest {
    var attributes: Array<AuthUserAttribute>
    var options: AuthUpdateUserAttributesRequest.Options?
    
    init(dict: NSMutableDictionary) {
        self.attributes = (dict["attributes"] as! Array<Dictionary<String, String>>).map {
            return createAuthUserAttribute(key: $0["userAttributeKey"]!, value: $0["value"]!);
        }
        self.options = createOptions(options: dict["options"] as! Dictionary<String, Any>?)
    }
    
    func createOptions(options: Dictionary<String, Any>?) -> AuthUpdateUserAttributesOperation.Request.Options {
      let pluginOptions = AWSUpdateUserAttributesOptions(
          metadata: options?["clientMetadata"] as? [String : String]
      )
      return AuthUpdateUserAttributesOperation.Request.Options(pluginOptions: pluginOptions)
    }
    
    static func validate(dict: NSMutableDictionary) throws {
        let validationErrorMessage = "UpdateUserAttributeRequest Request malformed."
        if (dict["attributes"] == nil || !(dict["attributes"] is Array<Dictionary<String, String>>)) {
            throw InvalidRequestError.auth(comment: validationErrorMessage, suggestion: String(format: ErrorMessages.missingAttribute, "attributes"))
        } else {
            let attributes = dict["attributes"] as! Array<Dictionary<String, Any>>;
            for attribute in attributes {
                try validateUserAttribute(attribute: attribute, validationErrorMessage: validationErrorMessage)
            }
        }
    }
}

File: packages/auth/amplify_auth_cognito_android/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/types/FlutterUpdateUserAttributesRequest.kt
/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amazonaws.amplify.amplify_auth_cognito.types

import com.amazonaws.amplify.amplify_auth_cognito.utils.createAuthUserAttribute
import com.amazonaws.amplify.amplify_auth_cognito.utils.validateUserAttribute
import com.amazonaws.amplify.amplify_core.exception.ExceptionMessages
import com.amazonaws.amplify.amplify_core.exception.InvalidRequestException
import com.amplifyframework.auth.AuthUserAttribute
import com.amplifyframework.auth.cognito.options.AWSCognitoAuthUpdateUserAttributesOptions

data class FlutterUpdateUserAttributesRequest(val map: HashMap<String, *>) {

    val attributes: List<AuthUserAttribute> = (map["attributes"] as List<HashMap<*, *>>)
        .map { createAuthUserAttribute(it["userAttributeKey"] as String, it["value"] as String) }

    val options: AWSCognitoAuthUpdateUserAttributesOptions = createOptions(map["options"] as HashMap<String, Any>?)

    private fun createOptions(rawOptions: HashMap<String, *>?): AWSCognitoAuthUpdateUserAttributesOptions {
        val optionsBuilder = AWSCognitoAuthUpdateUserAttributesOptions.builder()
        if (rawOptions?.get("clientMetadata") != null) {
            optionsBuilder.metadata(rawOptions["clientMetadata"] as HashMap<String, String>)
        }
        return optionsBuilder.build()
    }

    companion object {
        private const val validationErrorMessage: String = "UpdateUserAttributesRequest Request malformed."
        fun validate(req: HashMap<String, *>?) {
            if (req == null) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("request map"))
            } else if (!req.containsKey("attributes") || req["attributes"] !is List<*>) {
                throw InvalidRequestException(validationErrorMessage, ExceptionMessages.missingAttribute.format("attributes"))
            } else {
                val attributes = req["attributes"] as List<HashMap<*, *>>
                if (attributes.isEmpty()) {
                    throw InvalidRequestException(validationErrorMessage, "The request must have at least one attribute.")
                } else {
                    attributes.forEach { validateUserAttribute(it, validationErrorMessage) }
                }
            }
        }
    }
}
